### PR TITLE
More SQL-like Substring

### LINKED ISSUE #1634: New SUBSTRING function differs from other SQL languages.
Our `SUBSTRING` function has args `startIndex` and `endIndex` and they are zero indexed.

Just taken a look at MySQL, Postgres, SPark SQL and Oracle and all have `startIndex` and `length and are one-indexed.

This is very hard to change once 5.0 goes out the door. We should look to change now.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 
Switch `SUBSTRING` from `SUBSTRING(str, startIndex, endIndex)` with zero-based indexing to `SUBSTRING(str, pos, len)` with:
 * one-based indexing
 * returns null for any null parameters
 * ignores out of bound indexes, (i.e. treats them as being just inbound)
as used by other SQL providers.

Uses can set `ksql.functions.substring.legacy.args` to true, (either through server properties or session properties, to switch the implementation back to legacy mode, i.e. `SUBSTRING(str, startIndex, endIndex)` and error handling that throws a lot of exceptions.

Fixes #1634

Because I've extended the description of `SubString`, I've also improved the formatting of the output of the `DESCRIBE FUNCTION x;` in the CLI, so that long descriptions are correctly split across lines and properly indented.  New output looks like:

```
ksql> describe function substring;

Name        : SUBSTRING
Author      : Confluent
Overview    : Returns a substring of the passed in value.
              The behaviour of this function changed in release 5.1. It is 
              possible to switch the function back to pre-v5.1 functionality via
               the setting:
              	ksql.functions.substring.legacy.args
              This can be set globally, through the server configuration file, 
              or per sessions or query via the set command.
Type        : scalar
Jar         : internal
Variations  : 

  Variation   : SUBSTRING(value VARCHAR, pos INT, len INT)
  Returns     : VARCHAR
  Description : Returns a substring of str that starts at pos and is of length 
                len
  value       : The source string. If null, then function returns null.
  pos         : The base-one position the substring starts from. If null, then 
                function returns null. (If in legacy mode, this argument is 
                base-zero)
  len         : The length of the substring to extract. If null, then function 
                returns null. (If in legacy mode, this argument is the endIndex 
                (exclusive), rather than the length

  Variation   : SUBSTRING(value VARCHAR, pos INT)
  Returns     : VARCHAR
  Description : Returns a substring of str that starts at pos  and continues to 
                the end of the string
  value       : The source string. If null, then function returns null.
  pos         : The base-one position the substring starts from. If null, then 
                function returns null. (If in legacy mode, this argument is 
                base-zero)
```

Related release notes added in this PR: https://github.com/confluentinc/docs/pull/1095

### Testing done 
Unit tests added and JSON based functional test.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by dguy** [comment]
This breaks compatibility with 4.1.x substring function

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java [hunk 1]
```diff
@@ -21,21 +21,35 @@
...
+    description = "Returns a substring of the passed in value")
 public class Substring {
 
-  @Udf(description = "Returns a string that is a substring of this string. The"
-      + " substring begins with the character at the specified startIndex and"
-      + " extends to the end of this string.")
-  public String substring(final String value, final int startIndex) {
-    return value.substring(startIndex);
+  @Udf(description = "Returns a substring of str that starts at pos "
+                     + "and continues to the end of the string")
```
Comment (hjafarpour) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java]: Let's also document that the pos starts from 1 not 0.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java [hunk 2]
```diff
@@ -21,21 +21,35 @@
...
+    final int start = getStartIndex(str, pos);
+    return str.substring(start);
   }
 
-  @Udf(description = "Returns a string that is a substring of this string. The"
-      + " substring begins with the character at the specified startIndex and"
-      + " extends to the character at endIndex -1.")
-  public String substring(final String value, final int startIndex, final int endIndex) {
-    return value.substring(startIndex, endIndex);
+  @Udf(description = "Returns a substring of str that starts at pos and is of length len")
```
Comment (blueedgenick) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java]: was expecting to see some shiny new @UdfParameter decoration here to help teh user know which arg is the length and which one the startPos ? :-)

Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java]: Now on master branch - so annotations added!

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java [hunk 3]
```diff
@@ -16,35 +16,134 @@
...
+        + "This can be set globally, through the server configuration file, "
+        + "or per sessions or query via the set command.")
+public class Substring implements Configurable {
+
+  private Impl impl = new CurrentImpl();
+
+  @Override
+  public void configure(final Map<String, ?> props) {
+    final boolean legacyArgs =
+        getProps(props, KSQ_FUNCTIONS_PROPERTY_PREFIX + "substring.legacy.args", false);
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/string/Substring.java]: Fixed.

**Comment by big-andy-coates** [comment]
@rodesai thanks for the review! I've pushed some changes to address your comments.

**Pull request merged**